### ISSUE TITLE
Construction of Parameter_list and Rname without mapping.

### ISSUE THREAD

**Issue by d-winsor** [GabrielDosReis/ipr#41]
In PR #40 Fundecls were given a std::variant of Mapping or Parameter_list however the construction of Parameters and Rname are still tied to mappings.

```cpp
impl::Parameter* Mapping::param(const ipr::Name& n, const impl::Rname& rn);
impl::Rname* expr_factory::rname_for_next_param(const impl::Mapping& map, const ipr::Type& t);
impl::Parameter* Lexicon::make_parameter(const ipr::Name& n, const ipr::Type& t, impl::Mapping& m);
```

Some new functions in <ipr/impl> will be needed to construct Parameters without a mapping.

**Comment by d-winsor** [comment]
On that same note Parameter_list will need to be stored somewhere as it was previously stored as a member (by value) of Mapping.

**Comment by GabrielDosReis** [comment]
Yes.  What the intrinsic semantics call for is: (1) implementation for non-defining function declaration; and (2) implementation node for a function definition.

When done properly, (1) would have its own set of parameters, and (2) as usual will use a `Mapping` as initializer as usual.  The upshot is that an `ipr::Fundecl` has two concrete implementation nodes.  I think that is tripping up `ipr::impl::trait` which makes the assumption that each IPR interface node has exactly one implementation class.  That is not the case here.  The variant trick is trying to bring those views into the same unique projection.

Do we know why we got `Lexicon` have a `make_parameter()` in the first place?

**Comment by d-winsor** [comment]
I believe `make_parameter()` is just a shorthand for writing `mapping.param(name, *rname_for_next_param(mapping, type))`.

**Comment by GabrielDosReis** [comment]
> I believe make_parameter() is just a shorthand for writing mapping.param(name, *rname_for_next_param(mapping, type))

I think we want `make_parameter()` to be tied to something is parameterized: either a `Mapping` or a `Fundecl` (or a `Template`, but that is a named `Mapping`).  We shouldn't create a `Parameter` outside of those contexts.  So, I would recommend to remove it from `Lexicon`.

**Comment by GabrielDosReis** [comment]
Update:
With [PR 227](https://github.com/GabrielDosReis/ipr/pull/227), `Rname` is gone.  So, a `Parameter` node can now only be created by calling `impl::Parameter_list::add_member()` or `impl::Mapping::param()` by supplying a name and a type.  No more `Rname`.

Is your issue resolved?  What else is left regarding this?

**Comment by GabrielDosReis** [comment]
I consider this issue fixed, based on the recent additions.

Please reopen if there are new information.

**Comment by GorNishanov** [comment]
I believe the issue is still present. There is no mechanism to manufacture a `Parameter_list`.
Currently, impl::Parameter_list is embedded in impl::Lambda and impl::Mapping. There is no `make_parameter_list` factory function.

I believe this issue was born out of the need to represent a function declaration:

`void f(int = 5, double = 3.14);`

In this case, we don't have a Mapping to store a parameter list, and we added

```c++
struct fundecl_data : std::variant<impl::Parameter_list*, impl::Mapping*> {
    impl::Parameter_list* parameters() const { return std::get<0>(*this); }
    impl::Mapping* mapping() const { return std::get<1>(*this); }
};
```

To initialize this for a function declaration, we would need to manufacture a `Parameter_list`.

**Comment by GabrielDosReis** [comment]
Reopening after offline conversation with @GorNishanov : a non-defining declaration for a function still requires a separate `Parameter_list` in the current API set.
